File: src/db/queries/snapshots.ts
import { asc, desc, lte } from "drizzle-orm";
import { db, schema } from "@/db";

const { snapshots } = schema;

export type Snapshot = typeof snapshots.$inferSelect;

export function listSnapshots(): Snapshot[] {
  return db.select().from(snapshots).orderBy(asc(snapshots.date)).all();
}

export function getSnapshotOnOrBefore(date: string): Snapshot | null {
  return (
    db
      .select()
      .from(snapshots)
      .where(lte(snapshots.date, date))
      .orderBy(desc(snapshots.date))
      .limit(1)
      .get() ?? null
  );
}

export function upsertSnapshot(
  date: string,
  totalValueJpy: number,
  totalCostJpy: number,
  unrealizedPnlJpy: number,
): void {
  db
    .insert(snapshots)
    .values({ date, totalValueJpy, totalCostJpy, unrealizedPnlJpy })
    .onConflictDoUpdate({
      target: snapshots.date,
      set: { totalValueJpy, totalCostJpy, unrealizedPnlJpy },
    })
    .run();
}

File: OLD/src/db/snapshots.rs
use sqlx::{SqlitePool, FromRow};
use chrono::NaiveDate;

#[derive(Debug, FromRow, Clone)]
pub struct Snapshot {
    pub date: NaiveDate,
    pub total_value_jpy: f64,
    pub total_cost_jpy: f64,
    pub unrealized_pnl_jpy: f64,
}

pub async fn list_snapshots(pool: &SqlitePool) -> Result<Vec<Snapshot>, sqlx::Error> {
    sqlx::query_as::<_, Snapshot>(
        "SELECT date, total_value_jpy, total_cost_jpy, unrealized_pnl_jpy FROM snapshots ORDER BY date ASC"
    )
    .fetch_all(pool)
    .await
}

pub async fn get_snapshot_on_or_before(
    pool: &SqlitePool,
    date: NaiveDate,
) -> Result<Option<Snapshot>, sqlx::Error> {
    sqlx::query_as::<_, Snapshot>(
        "SELECT date, total_value_jpy, total_cost_jpy, unrealized_pnl_jpy
         FROM snapshots WHERE date <= ? ORDER BY date DESC LIMIT 1"
    )
    .bind(date)
    .fetch_optional(pool)
    .await
}

pub async fn upsert_snapshot(pool: &SqlitePool, date: NaiveDate, total_value: f64, total_cost: f64, pnl: f64) -> Result<(), sqlx::Error> {
    sqlx::query(
        r#"
        INSERT INTO snapshots (date, total_value_jpy, total_cost_jpy, unrealized_pnl_jpy)
        VALUES (?, ?, ?, ?)
        ON CONFLICT(date) DO UPDATE SET
            total_value_jpy = excluded.total_value_jpy,
            total_cost_jpy = excluded.total_cost_jpy,
            unrealized_pnl_jpy = excluded.unrealized_pnl_jpy
        "#
    )
    .bind(date)
    .bind(total_value)
    .bind(total_cost)
    .bind(pnl)
    .execute(pool)
    .await?;
    Ok(())
}
